main.js

```diff
@@ -32,90 +148,44 @@ async function run() {
       core.startGroup('Downloading MSYS2...');
-      base = '%~dp0';
       let inst_dest = path.join(tmp_dir, 'base.exe');
-      await tc.downloadTool(inst_url, inst_dest);
-
-      let inst_checksum = '';
-      await exec.exec(`powershell.exe`, [`(Get-FileHash ${inst_dest} -Algorithm SHA256)[0].Hash`], {listeners: {stdout: (data) => { inst_checksum += data.toString(); }}});
-      if (inst_checksum.slice(0, -2).toUpperCase() !== checksum.toUpperCase()) {
-        core.setFailed(`The SHA256 of the installer does not match! expected ${checksum} got ${inst_checksum}`);
-        return;
-      }
+      await downloadInstaller(inst_dest);
 
       changeGroup('Extracting MSYS2...');
       await exec.exec(inst_dest, ['-y'], {cwd: dest});
+      msysRootDir = path.join(dest, 'msys64');
 
       changeGroup('Disable Key Refresh...');
-      let post_file = `${dest}\\msys64\\etc\\post-install\\07-pacman-key.post`;
-      await exec.exec(`powershell.exe`, [`((Get-Content -path ${post_file} -Raw) -replace '--refresh-keys', '--version') | Set-Content -Path ${post_file}`]);
+      await disableKeyRefresh(msysRootDir);
       core.endGroup();
+    } else {
+      msysRootDir = path.join('C:', 'msys64');
     }
 
-    let wrap = [
-      `@echo off`,
-      `setlocal`,
-      `IF NOT DEFINED MSYS2_PATH_TYPE set MSYS2_PATH_TYPE=` + p_pathtype,
-      `set CHERE_INVOKING=1`,
-      base + `\\msys64\\usr\\bin\\bash.exe -leo pipefail %*`
-    ].join('\r\n');
-
-    let cmd = path.join(dest, 'msys2.cmd');
-    fs.writeFileSync(cmd, wrap);
-
-    core.addPath(dest);
-    const pkgCachePath = (p_release ? dest : 'C:') + `\\msys64\\var\\cache\\pacman\\pkg\\`;
+    writeWrapper(msysRootDir, input, dest, 'msys2.cmd');
 
-    const msystem_allowed = ['MSYS', 'MINGW32', 'MINGW64'];
-    if (!msystem_allowed.includes(p_msystem.toUpperCase())) {
-      core.setFailed(`'msystem' needs to be one of ${ msystem_allowed.join(', ') }, got ${p_msystem}`);
-      return;
-    }
-    core.exportVariable('MSYSTEM', p_msystem.toUpperCase());
+    core.exportVariable('MSYSTEM', input.msystem);
 
-    // We want a cache key that is ideally always the same for the same kind of job.
-    // So that mingw32 and ming64 jobs, and jobs with different install packages have different caches.
-    let shasum = crypto.createHash('sha1');
-    shasum.update([p_release, p_update, p_pathtype, p_msystem, p_install].toString());
-    // We include "update" in the fallback key so that a job run with update=false never fetches
-    // a cache created with update=true. Because this would mean a newer version than needed is in the cache
-    // which would never be used but also would win during cache prunging because it is newer.
-    const baseCacheKey = 'msys2-pkgs-upd:' + p_update.toString();
-    const jobCacheKey = baseCacheKey + '-conf:' + shasum.digest('hex').slice(0, 8);
+    const packageCache = new PackageCache(msysRootDir, input);
```

Comment:
What do you think about moving lines 175-210 to a function? So that `run` is split into "installing" MSYS and "caching/starting/updating/adding packages".

Reply:
I'll have to restructure this anyway for the next caching changes. I'll try then.